cape-ph/cape-cod
# Formalize our usage of the Pulumi configuration

Currently we have many different methods/models of interacting with the Pulumi configuration and it is fragmented and hard to recover at a given point what is really going on in any arbitrary place in the configuration. This PR is an investigation into formalizing, centralizing, and improving how we handle configuration through the infrastructure.

### Plan

This is the _current_ plan that is up to changing at any moment through development and discussion, will be kept up to date.

- [x] Manage all interaction of the Pulumi configuration with a `util.config` library
- [x] Centralize the idea of the "configuration" rather than placing that information arbitrarily
  - Keep in mind that we want to maintain the ease of adding/removing pieces in the configuration
- [x] Make it so the configuration is interlaced with the defaults and "fully resolved" as step 1 in the infrastructure to make sure it gets the understanding of configuration finalized before the infrastructure starts
  - This is happening at the component level and each component is in charge of the configuration for themselves. If components want to access each others configurations, they should do so by accessing the `.config` variable from a component that is _created_ and _initialized_

## TO TEST

Run `pulumi preview` on `main` and this branch and see that there are no differences
 
Closes #69
